internal/tools/networkagent/main.go

```diff
@@ -0,0 +1,350 @@
+				EndpointName: endpoint.Name,
+			}
+			if _, err := s.client.AddNIC(ctx, nsReq); err != nil {
+				return nil, err
+			}
+			s.endpointToNicID[endpoint.Name] = nicID.String()
+		}
+
+	}
+
+	defer func() {
+		if err != nil {
+			_, _ = s.teardownHelper(ctx, req, containerNamespaceID)
+		}
+	}()
+
+	return &nodenetsvc.ConfigureNetworkingResponse{}, nil
+
+}
+
+func (s *service) teardownHelper(ctx context.Context, req *nodenetsvc.ConfigureNetworkingRequest, containerNamespaceID string) (*nodenetsvc.ConfigureNetworkingResponse, error) {
+	return s.teardownHNSHelper(ctx, req, containerNamespaceID)
+}
+
+func (s *service) teardownHNSHelper(ctx context.Context, req *nodenetsvc.ConfigureNetworkingRequest, containerNamespaceID string) (*nodenetsvc.ConfigureNetworkingResponse, error) {
+	eReq := &ncproxygrpc.GetEndpointsRequest{}
+	resp, err := s.client.GetEndpoints(ctx, eReq)
+	if err != nil {
+		return nil, err
+	}
+	for _, endpoint := range resp.Endpoints {
+		if endpoint.Namespace == containerNamespaceID {
+			nicID, ok := s.endpointToNicID[endpoint.Name]
+			if !ok {
+				log.G(ctx).WithField("name", endpoint.Name).Warn("endpoint was not assigned a NIC ID previously")
+				continue
+			}
+			// remove endpoints that are in the namespace as NICs
+			nsReq := &ncproxygrpc.DeleteNICRequest{
+				ContainerID:  req.ContainerID,
+				NicID:        nicID,
+				EndpointName: endpoint.Name,
+			}
+			if _, err := s.client.DeleteNIC(ctx, nsReq); err != nil {
+				log.G(ctx).WithField("name", endpoint.Name).Warn("failed to delete endpoint nic")
+			}
+			delete(s.endpointToNicID, endpoint.Name)
+		}
+	}
+	return &nodenetsvc.ConfigureNetworkingResponse{}, nil
+}
+
+func (s *service) ConfigureNetworking(ctx context.Context, req *nodenetsvc.ConfigureNetworkingRequest) (*nodenetsvc.ConfigureNetworkingResponse, error) {
+	log.G(ctx).WithField("req", req).Info("ConfigureNetworking request")
+
+	containerNamespaceID, ok := s.containerToNamespace[req.ContainerID]
+	if !ok {
+		return nil, fmt.Errorf("no namespace was previously created for containerID %s", req.ContainerID)
+	}
+
+	if req.RequestType == nodenetsvc.RequestType_Setup {
+		return s.addHelper(ctx, req, containerNamespaceID)
+	}
+	return s.teardownHelper(ctx, req, containerNamespaceID)
+}
+
+func (s *service) GetHostLocalIpAddress(ctx context.Context, req *nodenetsvc.GetHostLocalIpAddressRequest) (*nodenetsvc.GetHostLocalIpAddressResponse, error) {
+	return &nodenetsvc.GetHostLocalIpAddressResponse{IpAddr: ""}, nil
+}
+
+func (s *service) PingNodeNetworkService(ctx context.Context, req *nodenetsvc.PingNodeNetworkServiceRequest) (*nodenetsvc.PingNodeNetworkServiceResponse, error) {
+	return &nodenetsvc.PingNodeNetworkServiceResponse{}, nil
+}
+
+func main() {
+	var err error
+	ctx := context.Background()
+
+	flag.Parse()
+	conf, err := readConfig(*configPath)
+	if err != nil {
+		log.G(ctx).WithError(err).Fatalf("failed to read network agent's config file at %s", *configPath)
```

Comment:
Fatalf (and Fatal) should call `os.Exit(1)` internally so the `os.Exit` after is redundant (and will never get hit).